### PR TITLE
 [Qt] Adding Taskbar Hover Menu for Windows (Updated)

### PR DESCRIPTION
**Pull request by CDAGaming**
#2983 Updated with Typo Fixes and Compatibility with Latest Changes to origin/Master

## Required PRs (To Merge before This Can be Merged):
- [ ] https://github.com/citra-emu/ext-windows-bin/pull/12
## Credits
@Styleoshin for Original PR

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3143)

### CODE REVIEW

File: src/citra_qt/main.cpp [hunk 1]
```diff
@@ -278,6 +282,15 @@ void GMainWindow::InitializeHotkeys() {
     });
 }
 
+void GMainWindow::InitializeWindowsExtras() {
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: Don't think a "Windows" function should be even exposed on non windows platforms.

File: src/citra_qt/main.cpp [hunk 2]
```diff
@@ -734,10 +750,25 @@ void GMainWindow::OnPauseGame() {
     ui.action_Start->setEnabled(true);
     ui.action_Pause->setEnabled(false);
     ui.action_Stop->setEnabled(true);
+
+    UpdateWindowsExtras();
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: Is there a reason why these are being called on every major interaction?

File: src/citra_qt/main.cpp [hunk 3]
```diff
@@ -734,10 +750,25 @@ void GMainWindow::OnPauseGame() {
...
+}
+
+void GMainWindow::OnRestartGame() {
+    BootGame(current_game_path);
+}
+
+void GMainWindow::OnResumeGame() {
+    if (emu_thread->IsRunning()) {
+        OnPauseGame();
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: OnResume calls OnPause? Might want to comment why that is the case.

File: src/citra_qt/windowsextras.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    void ClickRestart();
+
+private:
+    QWinThumbnailToolBar* thumbbar = nullptr;
+    QWinThumbnailToolButton* play_pause = nullptr;
+    QWinThumbnailToolButton* stop = nullptr;
+    QWinThumbnailToolButton* restart = nullptr;
+
+    QWidget* parent = nullptr;
+};
```
Comment (j-selby) [inline src/citra_qt/windowsextras.h]: Newline

**Comment by j-selby** [comment]
This doesn't need any kind of CIA PR being mixed into it. Rebase please.

**Comment by CDAGaming** [comment]
Will Reopen after Rebasing on master

File: .gitmodules [hunk 5]
```diff
@@ -1,33 +1,33 @@
-[submodule "inih"]
-    path = externals/inih/inih
-    url = https://github.com/svn2github/inih
 [submodule "boost"]
-    path = externals/boost
```
Comment (j-selby) [inline .gitmodules]: What happened here? Why was stuff rearranged/tabbed?

File: .travis.yml [hunk 6]
```diff
@@ -24,7 +24,7 @@ matrix:
     - os: osx
       env: NAME="macos build"
       sudo: false
-      osx_image: xcode7.3
+      osx_image: xcode9.1
```
Comment (j-selby) [inline .travis.yml]: What does a taskbar menu have to do with the build system?

File: .travis/linux/deps.sh [hunk 7]
```diff
@@ -1,3 +1,10 @@
 #!/bin/sh -ex
 
 docker pull ubuntu:16.04
+sudo apt-get remove -y postgresql-9.1
```
Comment (j-selby) [inline .travis/linux/deps.sh]: Ditto.

File: .travis/macos/deps.sh [hunk 8]
```diff
@@ -1,4 +1,6 @@
 #!/bin/sh -ex
 
+rvm get head --auto-dotfiles
```
Comment (j-selby) [inline .travis/macos/deps.sh]: Ditto.

File: CMakeLists.txt [hunk 9]
```diff
@@ -166,9 +166,9 @@ if (NOT PNG_FOUND)
     message(STATUS "libpng not found. Some debugging features have been disabled.")
 endif()
 
-find_package(Boost 1.63.0 QUIET)
+find_package(Boost 1.65.1 QUIET)
```
Comment (j-selby) [inline CMakeLists.txt]: Why are you updating Boost/SDL? You didn't respond to before.

File: CMakeModules/DownloadExternals.cmake [hunk 10]
```diff
@@ -8,7 +8,7 @@ set(prefix "${CMAKE_BINARY_DIR}/externals/${lib_name}")
 if (NOT EXISTS "${prefix}")
     message(STATUS "Downloading binaries for ${lib_name}...")
     file(DOWNLOAD
-        https://github.com/citra-emu/ext-windows-bin/raw/master/${remote_path}${lib_name}.7z
+        https://github.com/CDAGaming/ext-windows-bin/raw/master/${remote_path}${lib_name}.7z
```
Comment (j-selby) [inline CMakeModules/DownloadExternals.cmake]: This shouldn't be changed.

File: src/citra_qt/configuration/configure_input.cpp [hunk 11]
```diff
@@ -31,23 +31,19 @@ static QString getKeyName(int key_code) {
     }
 }
 
-static void SetButtonKey(int key, Common::ParamPackage& button_param) {
-    button_param = Common::ParamPackage{InputCommon::GenerateKeyboardParam(key)};
-}
-
-static void SetAnalogKey(int key, Common::ParamPackage& analog_param,
-                         const std::string& button_name) {
+static void SetAnalogButton(const Common::ParamPackage& input_param,
```
Comment (j-selby) [inline src/citra_qt/configuration/configure_input.cpp]: Why is this stuff changed?

File: src/citra_qt/configuration/configure_input.h [hunk 12]
```diff
@@ -35,10 +37,11 @@ class ConfigureInput : public QWidget {
 private:
     std::unique_ptr<Ui::ConfigureInput> ui;
 
-    std::unique_ptr<QTimer> timer;
+    std::unique_ptr<QTimer> timeout_timer;
```
Comment (j-selby) [inline src/citra_qt/configuration/configure_input.h]: Ditto.

File: src/citra_qt/configuration/configure_input.ui [hunk 13]
```diff
@@ -289,6 +289,13 @@
         <bool>false</bool>
        </property>
        <layout class="QGridLayout" name="gridLayout_4">
+        <item row="2" column="0" colspan="2">
```
Comment (j-selby) [inline src/citra_qt/configuration/configure_input.ui]: Ditto.

File: src/citra_qt/main.cpp [hunk 14]
```diff
@@ -158,6 +168,10 @@ void GMainWindow::InitializeWidgets() {
     message_label->setAlignment(Qt::AlignLeft);
     statusBar()->addPermanentWidget(message_label, 1);
 
+    progress_bar = new QProgressBar();
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: Progress bar?

File: src/citra_qt/main.cpp [hunk 15]
```diff
@@ -696,6 +725,61 @@ void GMainWindow::OnMenuSelectGameListRoot() {
     }
 }
 
+void GMainWindow::OnMenuInstallCIA() {
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: Unrelated PR. Please rebase.

**Comment by CDAGaming** [comment]
Fully Rebased, now back to One Commit (Newline also Added to windowsextras.cpp)

File: src/citra_qt/main.cpp [hunk 16]
```diff
@@ -30,6 +30,9 @@
 #include "citra_qt/hotkeys.h"
 #include "citra_qt/main.h"
 #include "citra_qt/ui_settings.h"
+#ifdef _WIN64
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: This import should be put somewhere else.

File: src/citra_qt/main.h [hunk 17]
```diff
@@ -180,6 +188,9 @@ private slots:
 
     QAction* actions_recent_files[max_recent_files_item];
 
+    // Native Windows
```
Comment (j-selby) [inline src/citra_qt/main.h]: Not a particularly useful comment.

File: src/citra_qt/windowsextras.cpp [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    connect(stop, &QWinThumbnailToolButton::clicked, this, &WindowsExtras::ClickStop);
+    connect(restart, &QWinThumbnailToolButton::clicked, this, &WindowsExtras::ClickRestart);
+}
+
+void WindowsExtras::Show() {
+    thumbbar->setWindow(parent->windowHandle());
+}
+
+void WindowsExtras::UpdatePlay() {
+    play_pause->setEnabled(true);
```
Comment (j-selby) [inline src/citra_qt/windowsextras.cpp]: I asked this before, but I'm not sure I got a response. Where did these icons come from? Where they made by you, or are they license compatible?

File: src/citra_qt/main.cpp [hunk 19]
```diff
@@ -1003,6 +1048,12 @@ void GMainWindow::UpdateUITheme() {
     }
 }
 
+void GMainWindow::ShowWindowsExtras() {
```
Comment (j-selby) [inline src/citra_qt/main.cpp]: Repeat from last review.

Don't think a method called ShowWindowsExtras should be exposed on non-Windows platforms.

File: src/citra_qt/windowsextras.h [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+class WindowsExtras : public QObject {
+    Q_OBJECT
+
+public:
+    void Show();
+    void UpdatePlay();
+    void UpdatePause();
+    void UpdateStop();
+
+    WindowsExtras(QWidget* parent);
```
Comment (lioncash) [inline src/citra_qt/windowsextras.h]: Constructors should be the first thing in the public interface. This should also be:

```cpp
explicit WindowsExtras(QWidget* parent);
```

File: src/citra_qt/windowsextras.h [hunk 21]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QWidget>
+#include <QWinThumbnailToolBar>
```
Comment (lioncash) [inline src/citra_qt/windowsextras.h]: This include and the one below it can be forward declarations

File: src/citra_qt/windowsextras.cpp [hunk 22]
```diff
@@ -0,0 +1,59 @@
+#include "windowsextras.h"
+
+WindowsExtras::WindowsExtras(QWidget* parent) : parent(parent) {
+    InitializeThumbnailToolbar();
+}
+
+WindowsExtras::~WindowsExtras() {}
```
Comment (lioncash) [inline src/citra_qt/windowsextras.cpp]:
```cpp
WindowsExtras::~WindowsExtras() = default;
```

File: src/citra_qt/windowsextras.cpp [hunk 23]
```diff
@@ -0,0 +1,59 @@
+#include "windowsextras.h"
```
Comment (lioncash) [inline src/citra_qt/windowsextras.cpp]: This should be a relative include

```cpp
"citra_qt/windowsextras.h"
```

File: src/citra_qt/main.h [hunk 24]
```diff
@@ -190,6 +198,9 @@ private slots:
 
     QAction* actions_recent_files[max_recent_files_item];
 
+    // Windows Specific Functionality (Used for Thumbnail Toolbar)
+    WindowsExtras* windows_extras;
```
Comment (lioncash) [inline src/citra_qt/main.h]: This should likely not even be present on a non-windows system

File: src/citra_qt/main.cpp [hunk 25]
```diff
@@ -903,6 +934,20 @@ void GMainWindow::UpdateStatusBar() {
     emu_frametime_label->setVisible(true);
 }
 
+void GMainWindow::UpdateWindowsExtras() {
+#ifdef _WIN64
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: `_WIN32` should be used over `_WIN64`, there's nothing about the contained code that specifically relies on the OS being 64-bit. `_WIN32` is available regardless of whether the windows version is 32-bit or 64-bit. Ditto for other occurrences as well.

File: src/citra_qt/main.h [hunk 26]
```diff
@@ -152,6 +157,9 @@ private slots:
 
 private:
     void UpdateStatusBar();
+    void UpdateWindowsExtras();
+
+    QString current_game_path; // To restart the game
```
Comment (lioncash) [inline src/citra_qt/main.h]: This is a pretty vague comment that doesn't really explain any context.

File: src/citra_qt/main.cpp (hunk 16 above)
Comment (lioncash) [reply src/citra_qt/main.cpp]: It should also be relative to `citra_qt/` like the rest of them.

**Comment by CDAGaming** [comment]
Thank You for the Review lioncash, Will definitely adjust this PR tomorrow accordingly (Too late today to do so unfortunately)

File: src/citra_qt/windowsextras.cpp (hunk 18 above)
Comment (CDAGaming) [reply src/citra_qt/windowsextras.cpp]: The Icons should be perfectly license compatible.
(They were originally made by the Origin Poster of this PR)

Comment (CDAGaming) [reply src/citra_qt/windowsextras.cpp]: New Icons Added and Are GPL Compatible :)

**Pull request closed without merging**